File: Windows/UnicodeGenerator/UnicodeGenerator.cpp
#include <iostream>
#include <unordered_set>
#include <set>
#include <map>

#include <PGE/String/Key.h>
#include <PGE/File/TextWriter.h>
#include <PGE/File/TextReader.h>

using namespace PGE;

static const String INDENT = "    ";

// TODO: This desperately needs refactoring.
class SwitchWriter {
    protected:
        using Cases = std::set<String::OrderedKey>;
        using StatementMap = std::map<String::OrderedKey, Cases>;

        SwitchWriter(const String& name, const String& param, const String& preAction, const String& postAction, const String& retType = "void")
            : name(name), param(param), preAction(preAction), postAction(postAction), retType(retType) { }

        void write(TextWriter& out, const StatementMap& done) const {
            out.writeLine(retType + " Unicode::" + name + "(" + (!param.isEmpty() ? (param + ", ") : "") + "char16 ch) {");
            out.writeLine(INDENT + "switch (ch) {");
            for (const auto& it : done) {
                std::vector<String> chars = it.first.str.split(" ", true);
                bool first = true;
                for (const auto& sit : it.second) {
                    if (first) {
                        first = false;
                    } else {
                        out.writeLine(" [[fallthrough]];");
                    }
                    out.write(INDENT.multiply(2) + "case " + rawToChar(sit.str) + ":");
                }
                out.writeLine(" {");
                for (const String& toChr : chars) {
                    writeAction(out, rawToChar(toChr));
                }
                writeBreak(out);
            }
            out.writeLine(INDENT.multiply(2) + "default: {");
            writeDefaultAction(out);
            writeBreak(out);
            out.writeLine(INDENT + "}");
            out.writeLine("}");
        }

        virtual void writeAction(TextWriter& out, const String& item) const {
            out.writeLine(INDENT.multiply(3) + preAction + item + postAction);
        }

        virtual void writeDefaultAction(TextWriter& out) const {
            writeAction(out, "ch");
        }

    private:
        const String name;
        const String param;
        const String preAction;
        const String postAction;
        const String retType;

        static void writeBreak(TextWriter& out) {
            out.writeLine(INDENT.multiply(2) + "} break;");
        }

        static String rawToChar(const String& str) {
            return "u'\\u" + str + "'";
        }
};

// Both SwitchWriter implementations constitute degenerate builders.
class Folder : private SwitchWriter {
    public:
        Folder(const String& name)
            : SwitchWriter(name, "std::queue<char16>& queue", "queue.push(", ");") { }

        void feed(const String& from, const String& to, bool dominant) {
            // It dominantes!
            if (dominant) {
                pre[from] = to;
            } else {
                // We haven't mapped this yet!
                if (pre.find(from) == pre.end()) {
                    pre.emplace(from, to);
                }
            }
        }

        void write(TextWriter& out) const {
            StatementMap done;
            for (const auto& it : pre) {
                done[it.second.str].emplace(it.first.str);
            }

            SwitchWriter::write(out, done);
        }

    private:
        std::unordered_map<String::RedundantKey, String::RedundantKey> pre;
};

class Caser : private SwitchWriter {
    public:
        Caser(const String& name)
            : SwitchWriter(name, "String& str", "str += ", ";") { }

        void feed(const String& from, const String& to) {
            // First come, first serve.
            if (hasHad.find(from) == hasHad.end()) {
                done[to].emplace(from);
                hasHad.emplace(from);
            }
        }

        void write(TextWriter& out) const {
            SwitchWriter::write(out, done);
        }

    private:
        StatementMap done;
        std::unordered_set<String::Key> hasHad;
};

class Identifier : private SwitchWriter {
    public:
        Identifier(const String& name)
            : SwitchWriter(name, String(), String(), String(), "bool") { }

        void feed(const String& match) {
            matchingChars.emplace(match);
        }

        void write(TextWriter& out) const {
            StatementMap map;
            map.emplace("TODO", matchingChars);
            SwitchWriter::write(out, map);
        }

    private:
        Cases matchingChars;

        void writeAction(TextWriter& out, const String& item) const override {
            out.writeLine(INDENT.multiply(3) + "return true;");
        }

        void writeDefaultAction(TextWriter& out) const override {
            out.writeLine(INDENT.multiply(3) + "return false;");
        }
};

namespace UnicodeParser {
    using Consumer = std::function<void(const std::vector<String>&)>;

    void parse(const FilePath& path, const Consumer& func) {
        TextReader in(path);
        static String line; // Fuck it, static line cache. (Remove this if used elsewhere, thx.)
        while (!in.endOfFile()) {
            line = String();
            in.readLine(line);
            if (line.byteLength() == 0 || *line.charAt(0) == L'#') { continue; }
            std::vector<String> params = line.split(";", false);
            params[0] = params[0].trim();
            // We reached UTF-32 range!
            if (params[0].byteLength() == 5) { break; }
            func(params);
        }
    }
};

int main() {
    TextWriter out(FilePath::fromStr("../../Src/String/Unicode.cpp"));

    // Incase it wasn't obvious, this doesn't apply to this file.
    out.writeLine(
R"(
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//                                                                                                                                   //
//       ###    ##     ## ########  #######   ######   ######## ##    ## ######## ########     ###    ######## ######## ########     //
//      ## ##   ##     ##    ##    ##     ## ##    ##  ##       ###   ## ##       ##     ##   ## ##      ##    ##       ##     ##    //
//     ##   ##  ##     ##    ##    ##     ## ##        ##       ####  ## ##       ##     ##  ##   ##     ##    ##       ##     ##    //
//    ##     ## ##     ##    ##    ##     ## ##   #### ######   ## ## ## ######   ########  ##     ##    ##    ######   ##     ##    //
//    ######### ##     ##    ##    ##     ## ##    ##  ##       ##  #### ##       ##   ##   #########    ##    ##       ##     ##    //
//    ##     ## ##     ##    ##    ##     ## ##    ##  ##       ##   ### ##       ##    ##  ##     ##    ##    ##       ##     ##    //
//    ##     ##  #######     ##     #######   ######   ######## ##    ## ######## ##     ## ##     ##    ##    ######## ########     //
//                                                                                                                                   //
//    ######## #### ##       ########  //  ########   #######     ##    ##  #######  ########    ######## ########  #### ########    //
//    ##        ##  ##       ##        //  ##     ## ##     ##    ###   ## ##     ##    ##       ##       ##     ##  ##     ##       //
//    ##        ##  ##       ##        //  ##     ## ##     ##    ####  ## ##     ##    ##       ##       ##     ##  ##     ##       //
//    ######    ##  ##       ######    //  ##     ## ##     ##    ## ## ## ##     ##    ##       ######   ##     ##  ##     ##       //
//    ##        ##  ##       ##        //  ##     ## ##     ##    ##  #### ##     ##    ##       ##       ##     ##  ##     ##       //
//    ##        ##  ##       ##        //  ##     ## ##     ##    ##   ### ##     ##    ##       ##       ##     ##  ##     ##       //
//    ##       #### ######## ########  //  ########   #######     ##    ##  #######     ##       ######## ########  ####    ##       //
//                                                                                                                                   //
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
)" + 1 // + 1 to skip leading newline.
    );
    out.writeLine("#include <PGE/String/Unicode.h>");
    out.writeLine("#include \"UnicodeInternal.h\"");
    out.writeLine();
    out.writeLine("using namespace PGE;");
    out.writeLine();

    Folder foldedSwitch("fold");
    UnicodeParser::parse(FilePath::fromStr("CaseFolding.txt"), [&](const std::vector<String>& params) {
        char16 type = *params[1].trim().charAt(0);
        // Full/Multi-char folding dominates!
        foldedSwitch.feed(params[0], params[2], type == L'F');
    });
    foldedSwitch.write(out);

    out.writeLine();

    Caser upSwitch("up");
    Caser downSwitch("down");

    // MANUAL OVERRIDE
    upSwitch.feed("00DF", "1E9E");
    //

    Identifier spaceIdent("isSpace");

    UnicodeParser::parse(FilePath::fromStr("SpecialCasing.txt"), [&](const std::vector<String>& params) {
        // We don't want any conditional mappings.
        if (*params[4].trim().charAt(0) != L'#') { return; }
        if (params[3].byteLength() != 0) {
            upSwitch.feed(params[0], params[3]);
        }
        if (params[1].byteLength() != 0) {
            downSwitch.feed(params[0], params[1]);
        }
    });

    UnicodeParser::parse(FilePath::fromStr("UnicodeData.txt"), [&](const std::vector<String>& params) {
        if (params[12].byteLength() != 0) {
            upSwitch.feed(params[0], params[12]);
        }
        if (params[13].byteLength() != 0) {
            downSwitch.feed(params[0], params[13]);
        }

        if (*params[2].charAt(0) == L'Z' || params[2] == "Cc") {
            spaceIdent.feed(params[0]);
        }
    });

    upSwitch.write(out);

    out.writeLine();

    downSwitch.write(out);

    out.writeLine();

    spaceIdent.write(out);

    std::cout << "Generated Unicode code!" << std::endl;
    return 0;
}

File: .github/workflows/unicode-generator.js
const fs = require("fs");
const cp = require("child_process");

if (process.argv.length !== 3) {
    console.log("Wrong command line arguments");
    return;
}
const fileName = process.argv[2];

const header = fs.readFileSync(`${__dirname}/unicode-header.txt`);

const file = {
    contents: header,
    write: function(cnt) {
        this.contents += cnt;
    },
    writeLine: function(cnt = "") { this.write(cnt + "\n") },
    writeSwitch: function(map, name, param, action, retType = "void", defaultAction = () => action("ch")) {
        const strCmp = (a, b) => a < b ? -1 : (b > a ? 1 : 0);
        const codePointToCharLiteral = (codepoint) => `u'\\u${codepoint}'`;

        let arr = [];
        for (const key in map) {
            arr.push({ key: key, value: map[key] });
        }
        arr.sort((a, b) => strCmp(a.value[0], b.value[0]));

        const INDENT = "    ";
        const optionalParam = param.length !== 0 ? (param + ", ") : "";
        this.writeLine(`${retType} Unicode::${name}(${optionalParam}char16 ch) {`);
        this.writeLine(`${INDENT}switch (ch) {`);
        for (const mapping of arr) {
            let first = true;
            for (const mappedChar of mapping.value.sort(strCmp)) {
                if (first) { first = false; }
                else { this.writeLine(" [[fallthrough]];"); }
                this.write(`${INDENT.repeat(2)}case ${codePointToCharLiteral(mappedChar)}:`);
            }
            this.writeLine(" {");
            for (const mapped of mapping.key.split(" ")) {
                this.writeLine(INDENT.repeat(3) + action(codePointToCharLiteral(mapped)));
            }
            this.writeLine(`${INDENT.repeat(2)}} break;`);
        }
        this.writeLine(`${INDENT.repeat(2)}default: {`);
        this.writeLine(INDENT.repeat(3) + defaultAction());
        this.writeLine(`${INDENT.repeat(2)}} break;`);
    
        this.writeLine(INDENT + "}");
        this.writeLine("}");
    }
}

class Caser {
    #cases = {};
    #hasHad = new Set();
    feed(from, to) {
        from = from.trim();
        to = to.trim();
        if (to.length === 0 || from === to) { return; }
        // First come, first served
        if (!this.#hasHad.has(from)) {
            this.#hasHad.add(from);
            this.#cases[to] = this.#cases[to] || [];
            this.#cases[to].push(from);
        }
    }
    write(file, name) {
        file.writeSwitch(this.#cases, name, "String& str", (ch) => `str += ${ch};`);
    }
}

let map = {};
parseUnicode("https://www.unicode.org/Public/UCD/latest/ucd/CaseFolding.txt", (params) => {
    // It's full/multi-char folding or not yet folded!
    params[0] = params[0].trim();
    params[2] = params[2].trim();
    if (params[0] !== params[2] && (params[1].trim() === "F" || !(params[0] in map))) {
        map[params[0]] = params[2];
    }
});
let flipMap = {};
for (const key in map) {
    flipMap[map[key]] = flipMap[map[key]] || [];
    flipMap[map[key]].push(key);
}
file.writeSwitch(flipMap, "fold", "CircularArray<char16>& queue", (ch) => `queue.pushBack(${ch});`);
file.writeLine();

const up = new Caser();
const down = new Caser();
// MANUAL OVERRIDE
up.feed("00DF", "1E9E");
//
parseUnicode("https://unicode.org/Public/UNIDATA/SpecialCasing.txt", (params) => {
    // We don't want any conditional mappings
    if (params[4].trim()[0] !== "#") { return; }
    up.feed(params[0], params[3]);
    down.feed(params[0], params[1]);
});

let spaceIdent = [];
parseUnicode("https://unicode.org/Public/UNIDATA/UnicodeData.txt", (params) => {
    up.feed(params[0], params[12]);
    down.feed(params[0], params[13]);
    if (params[2][0] === "Z" || params[2] === "Cc") {
        spaceIdent.push(params[0]);
    }
});

up.write(file, "up");
file.writeLine();
down.write(file, "down");
file.writeLine();
file.writeSwitch({spaceIdent}, "isSpace", "", () => "return true;", "bool", () => "return false;");
fs.writeFileSync(fileName, file.contents);

console.log("Generated Unicode code!")

function parseUnicode(url, func) {
    const unicode = cp.execSync(`curl ${url}`, { maxBuffer: 10_000_000 }).toString();
    for (const line of unicode.split("\n")) {
        if (line.length === 0 || line[0] === "#") { continue; }
        const params = line.split(";");
        // We reached UTF-32 range!
        if (params[0].length === 5) { break; }
        func(params);
    }
}
